### ISSUE TITLE
Getting parital derivative for fmus between enter_initialization_mode and exit_initialization_mode

### ISSUE THREAD

**Issue by arun3688** [OpenModelica/OpenModelica#8440]
### Description

Currently OpenModelica generated fmus support getting `partial derivatives` of fmu's using `fmi2GetDirectionalDerivative for the `simulation DAE`. It must be also possible to get the partial derivatives of fmus during `enter initialization mode` and `exit initialization mode`

### Example
oms_setCommandLineOption("--suppressPath=true")
oms_setTempDirectory("./getDirectionalDerivative_CS_lua/")

```
oms_newModel("model")
oms_addSystem("model.root", oms_system_wc)
oms_addSubModel("model.root.CauerLowPassAnalog", "Modelica.Electrical.Analog.Examples.CauerLowPassAnalog.fmu")
oms_setResultFile("model", "getDirectionalDerivatives_CS.mat", 10)
oms_instantiate("model")
-- getting partial derivative between enter and exit Initialization mode is not working
oms_getDirectionalDerivative("model.root.CauerLowPassAnalog.der(C1.v)")
oms_initialize("model")
print("info:    Partial Derivatives after Initialization")
print("info:      model.root.CauerLowPassAnalog.der(C1.v): " .. oms_getDirectionalDerivative("model.root.CauerLowPassAnalog.der(C1.v)"))
```
Save the above as a LUA Script and execute using OMSimulator 

`>>> OMSimulator getdirectionalDerivative.lua
`
It would be nice if we can generate a simple model and test the partial derivatives in `OMSimulator `to verify the results, 
Currently i verified the results using `FMPY` with fmus generated from `Dymola`. The `OpenModelica` generated `fmus ` also gives the same results as dymola, but only works after initialization. However the fmus generated from Dymola works well in `OMSimulator` in all stages

**Comment by AnHeuermann** [comment]
Here is an example similar to the one that is used in the FMI Standard (v2.0.3) 2.1.9 Getting Partial Derivatives:

```modelica
model A
  input Real u1(start=1);
  input Real u2(start=0);
  input Real u3(start=1);
  input Real u4(start=1);
  output Real y1;
  output Real y2;
  Real x;
initial equation
  x = 1;
equation
  der(x) = x*u1^2;
  /* g1 */ y1 = x + u1*u3 + u4;
  /* g2 */ y2 = x + u1^2;
end A;
```

The Jacobian for the simulation system:
![image](https://user-images.githubusercontent.com/38031952/150323433-4cf0c32e-1f9f-4dc2-a618-08721fdbff41.png)

@arun3688 is this enough info to construct a smaller test case?

**Comment by arun3688** [comment]
@AnHeuermann Thanks for the example, this should be good enough for testing

**Comment by kabdelhak** [comment]
> ```
> >>> OMSimulator simpleDerivative.lua
> info:    Partial Derivatives of OpenModelica fmus before Initialization
> info:      model.root.A.y1: 3.0
> info:      model.root.A.y2: 2.0
> info:      model.root.A.der(x): 0.0
> info:    Result file: A_CS.mat (bufferSize=10)
> info:    Partial Derivatives of OpenModelica fmus after Initialization
> info:      model.root.A.y1    : 4.0
> info:      model.root.A.y2    : 3.0
> info:      model.root.A.der(x): 3.0
> ```

The only way i am making sense of this if we see this as a jacobian for input and state variables.
x - states
dx - differentiated states
y - algebraic variables (+outputs)
i - inputs

the system gets solved like:
(dx,y) = F(x,i)

And then it gets partially differentiated for the states and inputs J=dF/dx,i
This is quite easy to do in this example since all equations are already causalized in the correct way, we just need to differentiate:

**equations**
 f1:  der(x) = x*u1^2;
 f2:  y1 = x + u1*u3 + u4;
 f3:  y2 = x + u1^2;

**Jacobian**
                        | x           |   u1       |  u2    |    u3    |  u4
      der(x)  f1   | 1*u1^2  | 2*x*u1  |  0      |  0        |   0
J = y1        f2   | 1           | u3         |  0      |  u1      |  1
      y2        f3   | 1           | 2*u1     |  0      |  0         |  0

If we now use the predefined start values (and initial equation for x) and sum up each row, we are getting the same values as reported (after initialization).

Does this sound correct? I can't make sense of it otherwise.

**Comment by kabdelhak** [comment]
The jacobian is a little messed up, but i think the idea comes across

**Comment by arun3688** [comment]
> If we now use the predefined start values (and initial equation for x) and sum up each row, we are getting the same values as reported (after initialization).
> 
> Does this sound correct? I can't make sense of it otherwise.

Yes that is correct, based on your jacobian matrix, and sum up each row, we get the correct results after initialization,

What about the results before initialization ?

**Comment by kabdelhak** [comment]
> What about the results before initialization ?

I can't really tell. For `y1` and `y2` it seems like the initial values have just been plugged into the corresponding equation, but that would not explain the value for `der(x)`.

By the way, this is the output from OMSimulator right? Do we know that this is correct? Do we have a specification on what we need to produce?

**Comment by arun3688** [comment]
> By the way, this is the output from OMSimulator right? Do we know that this is correct? Do we have a specification on what we need to produce?

Yes the results are from OMSimulator, Andreas can test the same with fmus generated from Dymola and we can verify the results

You can look into the fmi specification` chapter 2.1.9 `or `page 27`
[FMI-Specification-2.0.3.pdf](https://github.com/OpenModelica/OpenModelica/files/7911860/FMI-Specification-2.0.3.pdf)

**Comment by AnHeuermann** [comment]
Dymola FMU for `A` and simulated with FMPy:

```
Partial_derivatives before initialization
======================================== 
Partial_der of  y1 :=  [2.9999999999863465]    
Partial_der of  y2 :=  [1.9999999999964488]    
Partial_der of  der(x) :=  [1.9999999999964488]

Partial_derivatives after initialization
=======================================
Partial_der of  der(x) :=  [3.000000000397129]
Partial_der of  y1 :=  [3.999979999998876]
Partial_der of  y2 :=  [3.000000000003347]
```

**Comment by lochel** [comment]
I gues there should be one more unknown for the initialization system, right? 
And is this the sum of the jacobian rows?

**Comment by arun3688** [comment]
> I guess there should be one more unknown for the initialization system, right?
> And is this the sum of the jacobian rows?

Yes this is the sum of the jacobian rows, I will update the script to get all the column values

**Comment by kabdelhak** [comment]
> Dymola FMU for `A` and simulated with FMPy:
> 
> ```
> Partial_derivatives before initialization
> ======================================== 
> Partial_der of  y1 :=  [2.9999999999863465]    
> Partial_der of  y2 :=  [1.9999999999964488]    
> Partial_der of  der(x) :=  [1.9999999999964488]

Maybe what i thought is right. This looks like the initial values were just plugged into the equations, seems like we skip the computation of der(x) wheras dymola does it.

**Comment by arun3688** [comment]
@lochel here is the result of all individual column values of the jacobian, from `FMPY`

```
Partial_derivatives before initialization
========================================
der(x) ===> ['u1']
Partial_der of  der(x) / u1 :=  [0.0]

y1 ===> ['u1', 'u3', 'u4']
Partial_der of  y1 / u1 :=  [1.0]
Partial_der of  y1 / u3 :=  [1.0]
Partial_der of  y1 / u4 :=  [1.0]

y2 ===> ['u1']
Partial_der of  y2 / u1 :=  [2.0]

Partial_derivatives after initialization
=======================================
der(x) ===> ['x', 'u1']
Partial_der of  der(x) / x :=  [1.0]
Partial_der of  der(x) / u1 :=  [2.0]

y1 ===> ['x', 'u1', 'u3', 'u4']
Partial_der of  y1 / x :=  [1.0]
Partial_der of  y1 / u1 :=  [1.0]
Partial_der of  y1 / u3 :=  [1.0]
Partial_der of  y1 / u4 :=  [1.0]

y2 ===> ['x', 'u1']
Partial_der of  y2 / x :=  [1.0]
Partial_der of  y2 / u1 :=  [2.0]
```

**Comment by arun3688** [comment]
@lochel After doing `setReal` on the `inputs` during `enter_initialization_mode`, we get correct results for all the values of `initialization` and `simulation` **jacobian** matrix, see the results below

```
Partial_derivatives before initialization
========================================
x ===> [] ====> []

der(x) ===> ['u1'] ====> [[1.0]]
Partial_der of  der(x) w.r.t.  u1 :=  [2.0]

y1 ===> ['u1', 'u3', 'u4'] ====> [[1.0], [1.0], [1.0]]
Partial_der of  y1 w.r.t.  u1 :=  [1.0]
Partial_der of  y1 w.r.t.  u3 :=  [1.0]
Partial_der of  y1 w.r.t.  u4 :=  [1.0]

y2 ===> ['u1'] ====> [[1.0]]
Partial_der of  y2 w.r.t.  u1 :=  [2.0]

Partial_derivatives after initialization
=======================================
der(x) ===> ['x', 'u1'] ===> [[1.0], [1.0]]
Partial_der of  der(x) w.r.t.  x :=  [1.0]
Partial_der of  der(x) w.r.t.  u1 :=  [2.0]

y1 ===> ['x', 'u1', 'u3', 'u4'] ===> [[1.0], [1.0], [1.0], [1.0]]
Partial_der of  y1 w.r.t.  x :=  [1.0]
Partial_der of  y1 w.r.t.  u1 :=  [1.0]
Partial_der of  y1 w.r.t.  u3 :=  [1.0]
Partial_der of  y1 w.r.t.  u4 :=  [1.0]

y2 ===> ['x', 'u1'] ===> [[1.0], [1.0]]
Partial_der of  y2 w.r.t.  x :=  [1.0]
Partial_der of  y2 w.r.t.  u1 :=  [2.0]

```

**Comment by AnHeuermann** [comment]
Looks like this is already fixed by #8477 or am I missing something?
